File: programmers/2020 KAKAO BLIND RECRUITMENT/자물쇠와 열쇠(Level3).py
'''
    Code written on Sep 10, 2020
    출처: 프로그래머스 코딩 테스트 연습, https://programmers.co.kr/learn/challenges
    2020 KAKAO BLIND RECRUITMENT: 자물쇠와 열쇠(Level3)
    link: https://programmers.co.kr/learn/courses/30/lessons/60059
'''
def revolve(key):
    return list(zip(*key[::-1]))


def padding(lock, N, key, M):
    size = N + 2 * (M - 1)
    extended = [[0] * size for _ in range(size)]
    for row in range(N):
        for col in range(N):
            extended[M-1+row][M-1+col] = lock[row][col]
    return extended


def insert(extended, key, x, y, M):
    for row in range(M):
        for col in range(M):
            extended[x+row][y+col] += key[row][col]
    

def remove(extended, key, x, y, M):
    for row in range(M):
        for col in range(M):
            extended[x+row][y+col] -= key[row][col]


def is_unlock(extended, N, M):
    for row in range(N):
        for col in range(N):
            if extended[M-1+row][M-1+col] != 1:
                return False
    return True
    

def solution(key, lock):
    N, M = len(lock), len(key)
    if is_unlock(lock, N, 0):
        return True
    extended = padding(lock, N, key, M)
    for _ in range(4):
        for x in range(N+M-1):
            for y in range(N+M-1):
                insert(extended, key, x, y, M)
                if is_unlock(extended, N, M):
                    return True
                remove(extended, key, x, y, M)
        key = revolve(key)
    return False

File: programmers/2020 KAKAO BLIND RECRUITMENT/자물쇠와 열쇠(Level3).java
/*
    Code written on Oct 2, 2020
    출처: 프로그래머스 코딩 테스트 연습, https://programmers.co.kr/learn/challenges
    2020 KAKAO BLIND RECRUITMENT: 자물쇠와 열쇠(Level3)
    link: https://programmers.co.kr/learn/courses/30/lessons/60059
*/
class Solution {
    public boolean solution(int[][] key, int[][] lock) {
        int keySize = key.length;
        int lockSize = lock.length;
        int widenSize = key.length - 1;
        int[][] widenLock = new int[widenSize * 2 + lockSize][widenSize * 2 + lockSize];
        
        padding(lock, widenLock, lockSize, widenSize);
        
        for (int i = 0; i < 4; i++) {
            for (int row = 0; row < widenSize + lockSize; row++) {
                for (int col = 0; col < widenSize + lockSize; col++) {
                    insertKey(key, widenLock, keySize, row, col);
                    if (isUnlock(widenLock, lockSize, widenSize))
                       return true;
                    detatchKey(key, widenLock, keySize, row, col);
                }
            }
            key = revolveKey(key, keySize);
        }
        return false;
    }
    
    public void padding(int[][] lock, int[][] widenLock, int lockSize, int widenSize) {
        for (int i = widenSize; i < widenSize + lockSize; i++)
            for (int j = widenSize; j < widenSize + lockSize; j++)
                widenLock[i][j] = lock[i-widenSize][j-widenSize];
    }
        
    public void insertKey(int[][] key, int[][] widenLock, int keySize, int row, int col) {  
        for (int i = row; i < row + keySize; i++)
            for (int j = col; j < col + keySize; j++)
                widenLock[i][j] += key[i-row][j-col];
    }
    
    public void detatchKey(int[][] key, int[][] widenLock, int keySize, int row, int col) {
        for (int i = row; i < row + keySize; i++)
            for (int j = col; j < col + keySize; j++)
                widenLock[i][j] -= key[i-row][j-col];
    }
    
    public boolean isUnlock(int[][] widenLock, int lockSize, int widenSize) {
        for (int i = widenSize; i < widenSize + lockSize; i++)
            for (int j = widenSize; j < widenSize + lockSize; j++)
                if (widenLock[i][j] != 1)
                    return false;
        return true;
    }

    public int[][] revolveKey(int[][]key, int keySize) {
        int[][] revolvedKey = new int[keySize][keySize];

        for (int i = 0; i < keySize; i++)
            for (int j = 0; j < keySize; j++)
               revolvedKey[i][j] = key[j][keySize-1-i];
        return revolvedKey;
    }
}
